src/engine/renderer/tr_surface.cpp

```diff
@@ -1113,7 +1200,14 @@ static void Tess_SurfaceMD5( md5Surface_t *srf )
 
 			VectorCopy( position, tessVertex->xyz );
 
-			Vector2Copy( surfaceVertex->texCoords, tessVertex->texCoords );
+			if ( glConfig2.halfFloatVertexAvailable )
+			{
+				Vector2Copy( surfaceVertex->f16TexCoords, tessVertex->f16TexCoords );
+			}
+			else
+			{
+				halfToFloat2( surfaceVertex->f16TexCoords, tessVertex->texCoords );
+			}
```

Comment:
The original code doesn't seem to be doing any half-float conversions here? If it is, then these 2 branches should be switched.

Reply:
The model structs (md5, iqm…) were not modified, they all store half-floats, so we need to copy them as half floats in the final struct if half-float is supported, or convert them back to float if not.

Note that I renamed half half-float related stuff by prefixing it by `f16`, so the original `Vector2Copy( surfaceVertex->texCoords, tessVertex->texCoords );` code was copying half floats. We are now converting them to float if the hardware doesn't support half float vertexes.

This code is correct.

In the future I may submit a patch to do the reverse: to always store them as float in those model structs, and convert them to half float if supported in the last step, because the fallback CPU code seems to do a lot of useless half-to-float conversions for computing the animations, which is stupid since the fallback CPU code is run on the lowest end hardware… But such design change is not in topic of that PR.